src/System.Private.CoreLib/src/System/Threading/WaitSubsystem.HandleManager.Unix.cs

```diff
@@ -0,0 +1,125 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System.Collections.Generic;
+using System.Diagnostics;
+using System.Runtime.CompilerServices;
+
+namespace System.Threading
+{
+    internal static partial class WaitSubsystem
+    {
+        [EagerStaticClassConstruction] // used during lazy class construction
+        private static class HandleManager
+        {
+            private const int InitialCapacity = 4;
+
+            /// <summary>
+            /// Prevents reusing a deleted index until at least there are this many deleted indexes in the free list. The
+            /// purpose is to allow free handles to remain free for some time so that we can discover invalid attempts at using
+            /// free handles.
+            /// </summary>
+            private const int MinimumFreeIndexCountBeforeReuse = 64;
+
+            private static WaitableObject[] s_objects = new WaitableObject[InitialCapacity];
+            private static int s_count;
+            private static LowLevelQueue<int> s_freeIndexes;
```

Comment:
This raises a few questions:

> That's why I am not thrilled about adding another one for what seems pretty low value.

Why are basic data structures in CoreFX then, and not in CoreLib? Perhaps the reason the copies exist is simply because the originals are in a place where they can't be used considering the current separation of code between CoreLib and CoreFX. Maybe some uses of these data structures are not good uses of those data structures. I'm fine with that. But when there are good uses of these basic data structures, the only solution based on the current architecture is to copy the code to CoreLib and that does not seem like a sufficient solution. Another solution is to desperately find an alternative that is more complicated and effectively accomplishes the same thing, I'm certainly not convinced by that argument.

> I believe that something like handle manager is best written directly against metal and kept as simple possible.

I'm not sure that this motivation is simply due to the lack of the appropriate data structures being available already.

> From what I can tell, CoreCLR PAL handle manager has a simple free list

CoreCLR PAL handle manager makes no attempt to prevent recently deleted handles from being reused, and in my opinion, is by no means simple, at the very least, not simpler than the handle manager in this change, hence my resistance. Why not make it simpler, if nobody has cared before? And if there are reasons to care, I'd be happy to discuss.